dvc/repo/stage.py

```diff
@@ -99,24 +99,100 @@ def _collect_specific_target(
     return [], file, name
 
 
+def locked(f):
+    @wraps(f)
+    def wrapper(loader: "StageLoad", *args, **kwargs):
+        with loader.repo.lock_repo():
+            return f(loader, *args, **kwargs)
+
+    return wrapper
+
+
 class StageLoad:
     def __init__(self, repo: "Repo") -> None:
-        self.repo = repo
+        self.repo: "Repo" = repo
+
+    @locked
+    def add(
+        self,
+        single_stage: bool = False,
+        fname: str = None,
+        validate: bool = True,
+        force: bool = False,
+        update_lock: bool = False,
+        from_cli: bool = False,
+        **stage_data,
+    ):
+        scm = self.repo.scm
+        stage = self.create(
+            single_stage=single_stage,
+            fname=fname,
+            validate=validate,
+            force=force,
+            from_cli=from_cli,
+            **stage_data,
+        )
+
+        with scm.track_file_changes(config=self.repo.config):
+            stage.dump(update_lock=update_lock)
+            stage.ignore_outs()
 
-    def create_from_cli(
-        self, validate: bool = False, **kwargs
+        return stage
+
+    def create(
+        self,
+        single_stage: bool = False,
+        validate: bool = True,
+        fname: str = None,
+        force: bool = False,
+        from_cli: bool = False,
+        **stage_data,
     ) -> Union["Stage", "PipelineStage"]:
-        """Creates a stage from CLI args passed as kwargs.
+        """Creates a stage.
 
         Args:
+            single_stage: if true, the .dvc file based stage is created,
+                fname is required in that case
+            fname: name of the file to use, not used for dvc.yaml files
             validate: if true, the new created stage is checked against the
                 stages in the repo. Eg: graph correctness,
                 potential overwrites in dvc.yaml file (unless `force=True`).
+            force: ignores overwrites in dvc.yaml file
+            stage_data: Stage data to create from
+                (see create_stage and loads_from for more information)
         """
-        from dvc.stage.utils import create_stage_from_cli
+        from dvc.stage import PipelineStage, Stage, create_stage, restore_meta
+        from dvc.stage.exceptions import InvalidStageName
+        from dvc.stage.utils import (
+            is_valid_name,
+            prepare_file_path,
+            prepare_kwargs_from_cli,
+            validate_state,
+        )
+
+        if from_cli:
```

Comment:
I might be missing something here,  but this `from_cli` logic is pretty weird, leaking cli into this lowlevel stage creation. It only signals that some of the arguments might be in string format, right? Or is there some additional logic/behavior that we need to apply?

Reply:
Feels like this should either not have cli in mind (e.g. from the API standpoint, the user might try to pass strings as arguments and we need to deal with them here) or the cli itself should properly prepare all the arguments before passing to `stage.add/create()` API

Reply:
cli details already leak into `dvc.run()` as they are 1:1 mapping with `dvc run` command. The `stage.add()` and `stage.create()` are meant to be high level API similar to `dvc run`. 

So, to remove this logic from here, either we would have to consider `dvc run` as an exception for now, and duplicate this logic on cli for stage:add and the `dvc.run()`, or bring it here.

The most important one is the parsing of `params` that we need to do inside `dvc run`, for which if we choose to change API, we could be able to move to `cli` logic.

Reply:
To clarify, I'd like to use the `params` as dict/list in API/Tests as well in addition to the cli str format.

Reply:
I am merging this for now. Let's discuss it (or, should I open a new issue?), and I'll update it swiftly. Thanks.

Reply:
nvmd, moved parsing logic outside.

Reply:
Sorry, forgot to react. Makes sense in the current circumstances. :+1: